src/XAD/Hessian.hpp

```diff
@@ -22,79 +22,53 @@
-    xad::Tape<xad::FReal<T>> *tape)
+std::vector<std::vector<T>> computeHessian(
+    const std::vector<xad::AReal<xad::FReal<T>>> &vec,
+    std::function<xad::AReal<xad::FReal<T>>(std::vector<xad::AReal<xad::FReal<T>>> &)> foo,
+    xad::Tape<xad::FReal<T>> *tape = xad::Tape<xad::FReal<T>>::getActive())
 {
-    unsigned int domain(static_cast<unsigned int>(v.size()));
-    std::vector<std::vector<xad::AReal<xad::FReal<T>>>> matrix(
-        domain, std::vector<xad::AReal<xad::FReal<T>>>(domain, 0.0));
-    std::function<xad::AReal<xad::FReal<T>>(std::vector<xad::AReal<xad::FReal<T>>> &)> foo = func;
-
-    tape->registerInputs(v);
-
-    for (unsigned int i = 0; i < domain; i++)
-    {
-        derivative(value(v[i])) = 1.0;
-        tape->newRecording();
-
-        xad::AReal<xad::FReal<T>> y = foo(v);
-        tape->registerOutput(y);
-        value(derivative(y)) = 1.0;
-
-        tape->computeAdjoints();
-
-        for (unsigned int j = 0; j < domain; j++)
-        {
-            // std::cout << "d2y/dx" << i << "dx" << j << " = " << derivative(derivative(v[j]))
-            //           << "\n";
-            matrix[i][j] = derivative(derivative(v[j]));
-        }
-
-        derivative(value(v[i])) = 0.0;
-    }
-
+    std::vector<std::vector<T>> matrix(vec.size(), std::vector<T>(vec.size(), 0.0));
+    computeHessian(vec, foo, begin(matrix), end(matrix), tape);
     return matrix;
 }
 
 // fwd_adj iterator
 template <class RowIterator, typename T>
 void computeHessian(
-    std::vector<xad::AReal<xad::FReal<T>>> &v,
-    std::function<xad::AReal<xad::FReal<T>>(std::vector<xad::AReal<xad::FReal<T>>> &)> func,
-    xad::Tape<xad::FReal<T>> *tape, RowIterator first, RowIterator last)
+    const std::vector<xad::AReal<xad::FReal<T>>> &vec,
+    std::function<xad::AReal<xad::FReal<T>>(std::vector<xad::AReal<xad::FReal<T>>> &)> foo,
+    RowIterator first, RowIterator last,
+    xad::Tape<xad::FReal<T>> *tape = xad::Tape<xad::FReal<T>>::getActive())
 {
+    auto v(vec);
     unsigned int domain(static_cast<unsigned int>(v.size()));
-    std::function<xad::AReal<xad::FReal<T>>(std::vector<xad::AReal<xad::FReal<T>>> &)> foo = func;
 
     if (std::distance(first, last) != domain)
         throw OutOfRange("Iterator not allocated enough space");
-    static_assert(detail::has_begin<typename std::iterator_traits<RowIterator>::value_type>::value,
-                  "RowIterator must dereference to a type that implements a begin() method");
+    static_assert(
+        !xad::detail::has_begin<typename std::iterator_traits<RowIterator>::value_type>::value,
+        "RowIterator must dereference to a type that implements a begin() method");
+
+    if (!tape)
+    {
+        std::unique_ptr<xad::Tape<xad::FReal<T>>> t;
+        t = std::unique_ptr<xad::Tape<xad::FReal<T>>>(new xad::Tape<xad::FReal<T>>());
+        if (!t)
+            throw xad::NoTapeException();
+        tape = t.get();
+    }
```

Comment:
This will not work - if you have a test covering this, it will fail - make sure you add that test. The `unique_ptr` goes out of scope on line 71, meaning the tape will be destroyed. You're left with a dangling pointer to `tape`.

Also the check for `!t` is not necessary. `operator new` will never return `nullptr`. It throws `bad_alloc` if allocation fails, which is ok to pass through to the user. And if the tape constructor throws, that can also be passed on to the user.

Reply:
I had a test for that but it was poorly written. I fixed both the test and the dangling pointer issue.